### ISSUE TITLE
Boot failed on hosts with LVM thin/raid logical volumes

### ISSUE THREAD

**Issue by eshikhov** [siderolabs/talos#9300]
### Discussed in https://github.com/siderolabs/talos/discussions/9140

<div type='discussions-op-text'>

<sup>Originally posted by **eshikhov** August  9, 2024</sup>
Hi,

When using the lvm mirror and lvm thin pool functionality, the OS is loaded randomly. 

When setting up for the first time, everything is fine, the kernel modules (dm_raid and dm_thin_pool) are loaded with the config. Then the lvm mirror and lvm thin pool volumes are created and everything works well until the server is rebooted (and the reboot can be caused by various reasons, be it talosctl upgrade with the --reboot-mode powercycle option or just talosctl reboot --mode powercycle). After a restart, talos does not start, but constantly goes into cyclic restarts and at some time of restart it can boot successfully.

Screenshots of unsuccessful launches that I managed to take:

![screenshot_1](https://github.com/user-attachments/assets/d3e049d1-f81e-40ad-9a34-54138d76f6d4)

![screenshot_2](https://github.com/user-attachments/assets/24fb60eb-63b2-429d-8882-b6a6e929f80a)

![screenshot_3](https://github.com/user-attachments/assets/4fa99fec-5ed5-4e73-aa0d-495b0c9a89f0)

![screenshot_4](https://github.com/user-attachments/assets/72adf05c-49ff-4473-8c70-ae86289d5698)

![screenshot_5](https://github.com/user-attachments/assets/64c96f52-72cd-41b7-9b16-8a46ee64678a)

Dump debug information about the cluster:

[support.zip](https://github.com/user-attachments/files/16562249/support.zip)

Talos v1.7.6</div>

**Comment by vladimirfx** [comment]
We reproduced the issue on VM. Logs and VMs can be found here: https://drive.google.com/drive/folders/1aj4GUOjO_EqSFdz4GgAZXhpCJ75ggZOm

In production, we are in a situation where we cannot reboot or update servers that have LV with thin or raid1 type. After reboot, the server can't boot from 30 minutes to several hours.

Of course, all servers have the correct modules configured and LVs created in the past year (using LVM CSI).

It looks like some race in kernel module initialization.

As a workaround, can you provide a machine config option to not activate LVM at boot time? (because all known LVM CSI implementations are active LVs themselves)

**Comment by vladimirfx** [comment]
@mst1711 FYI

**Comment by smira** [comment]
Can you extract the relevant bits as text? What is the root cause that you're seeing? It's almost impossible to follow screenshots, and they are not searchable.

**Comment by vladimirfx** [comment]
> Can you extract the relevant bits as text? What is the root cause that you're seeing? It's almost impossible to follow screenshots, and they are not searchable.

We provide a shared Google Disk folder with details. (https://drive.google.com/drive/folders/1aj4GUOjO_EqSFdz4GgAZXhpCJ75ggZOm)

VM boot logs from that folder:
[failed-boot-serial-console-output.txt](https://github.com/user-attachments/files/16962274/failed-boot-serial-console-output.txt)
[success-boot-serial-console-output.txt](https://github.com/user-attachments/files/16962275/success-boot-serial-console-output.txt)

Unfortunately, we can't provide such logs from bare metal because lack of a serial console.

**Comment by eshikhov** [comment]
The problem was discovered when updating from version 1.6.7 to version 1.7.5.
I'll check it again on version 1.6 releases

**Comment by frezbo** [comment]
@eshikhov would you be able to test with a custom iso/installer?

**Comment by eshikhov** [comment]
Yes, I can try

**Comment by eshikhov** [comment]
I checked it on a custom image, the problem cannot be reproduced. After booting, LVM volumes are available. Everything works.

**Comment by vladimirfx** [comment]
@smira can the fix be backported to 1.7?

**Comment by smira** [comment]
> @smira can the fix be backported to 1.7?

I don't know, this seems to be bigger/breaking change if things go wrong in some setups.

**Comment by vladimirfx** [comment]
> > @smira can the fix be backported to 1.7?
> 
> I don't know, this seems to be bigger/breaking change if things go wrong in some setups.

I understand your concern: changing the boot sequence in the patch release is against the usual practice. 
But the fixed version correctly initializes LVM (using udev rules) in contrast to unfixed which makes OS unbootable. So it is not just chore work for dropping the legacy part of the boot sequence - that effectively fixes real-user critical bugs without user-visible behavior change for others.

Anyway - thanks again for the good OS and support.

**Comment by vladimirfx** [comment]
> Talos 1.8.0-beta.1 will be released next Monday, so you can use that.

It is safe to use 1.8 worker nodes with 1.7 controll-plane?

**Comment by smira** [comment]
> > Talos 1.8.0-beta.1 will be released next Monday, so you can use that.
> 
> It is safe to use 1.8 worker nodes with 1.7 controll-plane?

In general no, but should be fine for this pair (1.7 <> 1.8).

**Comment by vladimirfx** [comment]
Just tested 1.8.0-beta.1 :

1. Nodes are booted normally - ok
2. Talos do not activate LVM volumes - **warn**

Udevd rules do not activate LVM logical volumes independent of their type (linear, thin, raid, or whatever).

For our case, it is ok, but it seems it should be nodes in release notes at least. IMO volume actuation should be done by CSI driver for their controlled volumes.

**Comment by frezbo** [comment]
@eshikhov @vladimirfx could you test with this installer image: `ghcr.io/frezbo/installer:v1.8.0-alpha.2-26-g18daedb51-dirty` ?

**Comment by vladimirfx** [comment]
> @eshikhov @vladimirfx could you test with this installer image: `ghcr.io/frezbo/installer:v1.8.0-alpha.2-26-g18daedb51-dirty` ?

Of course, we will check it soon. Can you share what changes it contains?

**Comment by frezbo** [comment]
> > @eshikhov @vladimirfx could you test with this installer image: `ghcr.io/frezbo/installer:v1.8.0-alpha.2-26-g18daedb51-dirty` ?
> 
> Of course, we will check it soon. Can you share what changes it contains?

yes, restructured code to behave like a normal os, if all volume groups is available and disks backing it is available it gets activated (with events from udev). The previous implementation tried to enable every volume group unconditionally. Following this reference: https://man7.org/linux/man-pages/man7/lvmautoactivation.7.html

**Comment by vladimirfx** [comment]
> yes, restructured code to behave like a normal os, if all volume groups is available and disks backing it is available it gets activated (with events from udev). The previous implementation tried to enable every volume group unconditionally. Following this reference: https://man7.org/linux/man-pages/man7/lvmautoactivation.7.html

Does it fail the boot in case of LVM activation failure?
In our case activation failed because of the kernel modules loading race (not device drivers but pure software components), so all PVs are available on boot and VGs start activating. I am scared that the new implementaion will fail the same way as the original...

**Comment by frezbo** [comment]
> > yes, restructured code to behave like a normal os, if all volume groups is available and disks backing it is available it gets activated (with events from udev). The previous implementation tried to enable every volume group unconditionally. Following this reference: https://man7.org/linux/man-pages/man7/lvmautoactivation.7.html
> 
> Does it fail the boot in case of LVM activation failure? In our case activation failed because of the kernel modules loading race (not device drivers but pure software components), so all PVs are available on boot and VGs start activating. I am scared that the new implementaion will fail the same way as the original...

it doesn't, it's a controller, so it'll just retry, and if lvm reports it's not healthy it just ignores

**Comment by eshikhov** [comment]
Tested this installer image: ghcr.io/frezbo/installer:v1.8.0-alpha.2-26-g18daedb51-dirty. 

LVM activation worked once after updating with reboot. 
After rebooting 10 times, activation did not work.

**Comment by frezbo** [comment]
> Tested this installer image: ghcr.io/frezbo/installer:v1.8.0-alpha.2-26-g18daedb51-dirty.
> 
> LVM activation worked once after updating with reboot. After rebooting 10 times, activation did not work.

Thanks for getting back, identified some issues after adding tests, once tests pass, I'll share another image

**Comment by frezbo** [comment]
> > Tested this installer image: ghcr.io/frezbo/installer:v1.8.0-alpha.2-26-g18daedb51-dirty.
> > LVM activation worked once after updating with reboot. After rebooting 10 times, activation did not work.
> 
> Thanks for getting back, identified some issues after adding tests, once tests pass, I'll share another image

Here we go: ghcr.io/frezbo/installer:v1.8.0-alpha.2-29-g61c34dfea

**Comment by frezbo** [comment]
> LVM activation has never been successful on this image.
> 
> /Users/eshikhov/Works/projects/yodha/client-cluster-config/infra/dev/dmesg_after_reboot.log /Users/eshikhov/Works/projects/yodha/client-cluster-config/infra/dev/dmesg_after_upgrade.log

this only activates if `pvscan` reports it needs to be activated, if it finds anything offline it'll skip. You can probably attach this info:

```bash
talosctl -n <node> logs controller-runtime | grep -i "block.LVMActivationController"
```

**Comment by eshikhov** [comment]
/Users/eshikhov/Works/projects/yodha/client-cluster-config/infra/dev/controller-runtime.log

**Comment by eshikhov** [comment]
[controller-runtime.log](https://github.com/user-attachments/files/17072821/controller-runtime.log)

**Comment by vladimirfx** [comment]
> > [controller-runtime.log](https://github.com/user-attachments/files/17072821/controller-runtime.log)
> 
> seems like `pvscan` doesn't report it needs to be activated, probably some conditions required missing

I guess the reason is the same as in the original issue - at **the moment** when `pvscan` is called required modules (raid1, thinpool) are not loaded yet. 
LVM depends not only on device availability but also on kernel module state. For that reason `lvm vgchange -ay` (used in `activateLogicalVolumes`) loads modules if needed (which is broken in Talos - see original screenshots). 
Looks like the root issue is a broken kernel module loading mechanic.

**Comment by frezbo** [comment]
> > > [controller-runtime.log](https://github.com/user-attachments/files/17072821/controller-runtime.log)
> > 
> > 
> > seems like `pvscan` doesn't report it needs to be activated, probably some conditions required missing
> 
> I guess the reason is the same as in the original issue - at **the moment** when `pvscan` is called required modules (raid1, thinpool) are not loaded yet. LVM depends not only on device availability but also on kernel module state. For that reason `lvm vgchange -ay` (used in `activateLogicalVolumes`) loads modules if needed (which is broken in Talos - see original screenshots). Looks like the root issue is a broken kernel module loading mechanic.

ahh do you have an example on how these vg's are created? if so we could try adding a test with external modules

**Comment by vladimirfx** [comment]
> > ahh do you have an example on how these vg's are created? if so we could try adding a test with external modules
> 
> Most CSIs create mirrored volume like this: `lvcreate --mirrors=1 --type=raid1 -L 1G --nosync -n mirroredlv vg0`

Such LVs require `dm_raid` kernel module. That module is capable of all raid levels (0,1,4,6,10) and on load, it tries the most effective checksumming algo which takes some time.

**Comment by frezbo** [comment]
> > > ahh do you have an example on how these vg's are created? if so we could try adding a test with external modules
> > 
> > 
> > Most CSIs create mirrored volume like this: `lvcreate --mirrors=1 --type=raid1 -L 1G --nosync -n mirroredlv vg0`
> 
> Such LVs require `dm_raid` kernel module. That module is capable of all raid levels (0,1,4,6,10) and on load, it tries the most effective checksumming algo which takes some time.

thanks, we'll see if we can write a test case for this

**Comment by frezbo** [comment]
did you have the `dm_raid` module loaded via machineconfig?

**Comment by vladimirfx** [comment]
> did you have the `dm_raid` module loaded via machineconfig?

Yes of course. See related discussion.

**Comment by frezbo** [comment]
> > did you have the `dm_raid` module loaded via machineconfig?
> 
> Yes of course. See related discussion.

Thanks, I was able to reproduce the issue, another fix would be coming, this actually makes talos code more robust, thanks again for providing proper info.

**Comment by frezbo** [comment]
@vladimirfx could you test this installer, hopefully the last one: ghcr.io/frezbo/installer:v1.8.0-alpha.2-31-g9fa08e843-dirty

**Comment by eshikhov** [comment]
does not work on this image: ghcr.io/frezbo/installer:v1.8.0-alpha.2-31-g9fa08e843-dirty

**Comment by frezbo** [comment]
Hmm weird, added a test that needs `dm_raid ` and pvscan seems to load them

**Comment by eshikhov** [comment]
On the old release of lvm volumes were automatically activated

The file you requested from a non-working node:
[support.zip](https://github.com/user-attachments/files/17098828/support.zip)

**Comment by vladimirfx** [comment]
I think this issue can be closed in favor of #9365 because node boot up is fixed.
